src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DelayUnassignedAllocation.java

```diff
@@ -0,0 +1,307 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.cluster.routing.allocation.allocator;
+
+import com.carrotsearch.hppc.cursors.ObjectCursor;
+import com.carrotsearch.hppc.cursors.ObjectObjectCursor;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.collect.Lists;
+import org.apache.lucene.util.ArrayUtil;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.routing.MutableShardRouting;
+import org.elasticsearch.cluster.routing.ShardRouting;
+import org.elasticsearch.cluster.routing.allocation.RoutingAllocation;
+import org.elasticsearch.common.component.AbstractComponent;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.unit.TimeValue;
+
+import java.util.*;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * The delayed unassigned allocation allows to control delayed allocation of shards due to nodes leaving the
+ * cluster with the assumption that they might come back, and the opertunity to reduce cluster allocations if
+ * they do.
+ */
+public class DelayUnassignedAllocation extends AbstractComponent {
+
+    public static final String DELAY_ALLOCATION_DURATION = "cluster.routing.allocation.delay_unassigned_allocation.duration";
+    public static final String DELAY_ALLOCATION_NODE_KEY = "cluster.routing.allocation.delay_unassigned_allocation.node_key";
+
+    private final NodeKey nodeKey;
+    private volatile TimeValue duration;
+
+    private final Map<Object, DelayedAllocation> delayedAllocations = new HashMap<>();
+    private final Cache<String, DiscoveryNode> seenNodes = CacheBuilder.newBuilder().expireAfterWrite(5, TimeUnit.MINUTES).build();
+
+    public DelayUnassignedAllocation(Settings settings) {
+        super(settings);
+        this.duration = settings.getAsTime(DELAY_ALLOCATION_DURATION, TimeValue.timeValueMinutes(5));
+        this.nodeKey = NodeKey.valueOf(settings.get(DELAY_ALLOCATION_NODE_KEY, NodeKey.TRANSPORT_ADDRESS.toString()).toUpperCase(Locale.ROOT));
+        logger.debug("delayed allocation using duration [{}] and node_key [{}]", duration, nodeKey);
+    }
+
+    /**
+     * The duration to wait if a node leaves, a value below or equal to 0 means
+     * not taking node leaving into account at all (disabling delayed allocation).
+     */
+    public TimeValue getDuration() {
+        return this.duration;
+    }
+
+    /**
+     * The duration to wait if a node leaves, a value below or equal to 0 means
+     * not taking node leaving into account at all (disabling delayed allocation).
+     */
+    public void setDuration(TimeValue duration) {
+        this.duration = duration;
+    }
+
+    /**
+     * Clears that current delayed allocation.
+     */
+    public void clearDelayedAllocations() {
```

Comment:
doh!, thanks for spotting it, it used to be concurrent map, but then with the sync it was not needed and I missed it